otel-integration/k8s-helm/e2e-test/sinks.go

```diff
@@ -0,0 +1,198 @@
+	f := otlpreceiver.NewFactory()
+	cfg := f.CreateDefaultConfig().(*otlpreceiver.Config)
+	cfg.HTTP = nil
+	cfg.GRPC.NetAddr.Endpoint = "0.0.0.0:4317"
+
+	_, err := f.CreateMetrics(context.Background(), receivertest.NewNopSettings(), cfg, mc)
+	require.NoError(t, err, "failed creating metrics receiver")
+	rcvr, err := f.CreateTraces(context.Background(), receivertest.NewNopSettings(), cfg, tc)
+	require.NoError(t, err, "failed creating traces receiver")
+	require.NoError(t, rcvr.Start(context.Background(), componenttest.NewNopHost()))
+
+	return func() {
+		assert.NoError(t, rcvr.Shutdown(context.Background()))
+	}
+}
+
+func WaitForMetrics(t *testing.T, entriesNum int, mc *consumertest.MetricsSink) {
+	timeoutSeconds := 180 // 3 minutes
+	require.Eventuallyf(t, func() bool {
+		count := len(mc.AllMetrics())
+		t.Logf("Waiting for metrics: got %d/%d", count, entriesNum)
+		return count >= entriesNum
+	}, time.Duration(timeoutSeconds)*time.Second, 1*time.Second,
+		"failed to receive %d entries in %d seconds",
+		entriesNum, timeoutSeconds)
+}
+
+func WaitForTraces(t *testing.T, entriesNum int, tc *consumertest.TracesSink) {
+	timeoutSeconds := 180 // 3 minutes
+	require.Eventuallyf(t, func() bool {
+		count := len(tc.AllTraces())
+		t.Logf("Waiting for traces: got %d/%d", count, entriesNum)
+		return count >= entriesNum // Changed > to >=
+	}, time.Duration(timeoutSeconds)*time.Second, 1*time.Second,
+		"failed to receive %d entries in %d seconds",
+		entriesNum, timeoutSeconds)
+}
+
+func ScanTracesForAttributes(t *testing.T, ts *consumertest.TracesSink, expectedService string, kvs map[string]ExpectedValue, scopeSpanAttrs []map[string]ExpectedValue) {
+	for i := 0; i < len(ts.AllTraces()); i++ {
+		traces := ts.AllTraces()[i]
+		for i := 0; i < traces.ResourceSpans().Len(); i++ {
+			resource := traces.ResourceSpans().At(i).Resource()
+			service, exist := resource.Attributes().Get(ServiceNameAttribute)
+			assert.True(t, exist, "Resource does not have the 'service.name' attribute")
+			if service.AsString() != expectedService {
+				continue
+			}
+			assert.NoError(t, assertExpectedAttributes(resource.Attributes(), kvs))
+
+			if len(scopeSpanAttrs) == 0 {
+				return
+			}
+
+			assert.NotZero(t, traces.ResourceSpans().At(i).ScopeSpans().Len())
+			assert.NotZero(t, traces.ResourceSpans().At(i).ScopeSpans().At(0).Spans().Len())
+
+			scopeSpan := traces.ResourceSpans().At(i).ScopeSpans().At(0)
+
+			// look for matching spans containing the desired attributes
+			for _, spanAttrs := range scopeSpanAttrs {
+				var err error
+				for j := 0; j < scopeSpan.Spans().Len(); j++ {
+					err = assertExpectedAttributes(scopeSpan.Spans().At(j).Attributes(), spanAttrs)
+					if err == nil {
+						break
+					}
+				}
+				assert.NoError(t, err)
+			}
+
+			return
+		}
+	}
+	t.Fatalf("no spans found for service %s", expectedService)
+}
+
+func assertExpectedAttributes(attrs pcommon.Map, kvs map[string]ExpectedValue) error {
```

Comment:
I think this whole function could be simplified to something like this:

```go
var notFoundAttrs []string

attrs.Range(func(k string, v pcommon.Value) bool {
	if val, found := kvs[k]; found {
		switch val.Mode {
		case AttributeMatchTypeEqual:
			if val.Value == v.AsString() {
				foundAttrs[k] = true
			}
		case AttributeMatchTypeRegex:
			matched, _ := regexp.MatchString(val.Value, v.AsString())
			if matched {
				foundAttrs[k] = true
			}
		case AttributeMatchTypeExist:
			foundAttrs[k] = true
		}
	} else {
		notFoundAttrs = append(notFoundAttrs, k)
	}
	return true
})

// if something is missing, add a summary with an overview of the expected and actual attributes for easier troubleshooting
if len(notFoundAttrs) > 0 {
    expectedJson, _ := json.MarshalIndent(kvs, "", "  ")
	actualJson, _ := json.MarshalIndent(attrs.AsRaw(), "", "  ")
	err = errors.Join(err, fmt.Errorf("one or more attributes were not found.\nExpected attributes:\n %s \nActual attributes: \n%s", expectedJson, actualJson))
}

return err
```

Reply:
thanks, yeah less loops

Reply:
oh i'll change it back cause the  `fmt.Errorf("attribute '%v' not found", k)` is quite useful when debugging

Reply:
still needs a second loop to print out the unfounded attributes